pkg/gitprovider/bitbucketserver.go

```diff
@@ -0,0 +1,374 @@
+	// Refer this developer community comment: https://community.developer.atlassian.com/t/obtain-authorised-users-username-from-api/24422/2
+	res, err := bitbucketDCClient.DefaultApi.GetApplicationProperties()
+	if err != nil {
+		return nil, err
+	}
+
+	username := res.Header.Get("X-Ausername")
+	if username == "" {
+		return nil, fmt.Errorf("X-Ausername header is missing")
+	}
+
+	user, err := bitbucketDCClient.DefaultApi.GetUser(username)
+	if err != nil {
+		return nil, err
+	}
+
+	var userInfo bitbucketv1.User
+	err = mapstructure.Decode(user.Values, &userInfo)
+
+	if err != nil {
+		return nil, err
+	}
+
+	response := &GitUser{}
+	response.Id = fmt.Sprintf("%d", userInfo.ID)
+	response.Username = username
+	response.Name = userInfo.DisplayName
+
+	return response, nil
+}
+
+func (g *BitbucketServerGitProvider) GetLastCommitSha(staticContext *StaticGitContext) (string, error) {
+	client := g.getApiClient()
+
+	branch := ""
+	if staticContext.Branch != nil {
+		branch = *staticContext.Branch
+	}
+
+	bitbucketDCClient, ok := client.(*bitbucketv1.APIClient)
+	if !ok {
+		return "", fmt.Errorf("Invalid Bitbucket Data Center/Server client")
+	}
+
+	commits, err := bitbucketDCClient.DefaultApi.GetCommits(staticContext.Id, staticContext.Id, map[string]interface{}{
+		"until": branch,
+	})
+
+	if err != nil {
+		return "", err
+	}
+
+	if len(commits.Values) == 0 {
+		return "", fmt.Errorf("No commits found")
+	}
+
+	commitList, err := bitbucketv1.GetCommitsResponse(commits)
+	if err != nil {
+		return "", err
+	}
+
+	return commitList[0].ID, nil
+}
+
+func (g *BitbucketServerGitProvider) getPrContext(staticContext *StaticGitContext) (*StaticGitContext, error) {
+	if staticContext.PrNumber == nil {
+		return staticContext, nil
+	}
+
+	repo := *staticContext
+
+	client := g.getApiClient()
+
+	bitbucketDCClient, ok := client.(*bitbucketv1.APIClient)
+	if !ok {
+		return nil, fmt.Errorf("Invalid Bitbucket Data Center/Server client")
+	}
+
+	pr, err := bitbucketDCClient.DefaultApi.GetPullRequest(staticContext.Id, staticContext.Id, int(*staticContext.PrNumber))
+	if err != nil {
+		return nil, err
+	}
+
+	prInfo, err := bitbucketv1.GetPullRequestResponse(pr)
+	if err != nil {
+		return nil, err
+	}
+
+	repo.Owner = prInfo.FromRef.Repository.Owner.DisplayName
+	repo.Name = prInfo.FromRef.Repository.Slug
+	repo.Id = prInfo.FromRef.Repository.Slug
+	repo.Branch = &prInfo.FromRef.DisplayID
+
+	return &repo, nil
+}
+
+func (g *BitbucketServerGitProvider) parseStaticGitContext(repoUrl string) (*StaticGitContext, error) {
+	staticContext, err := g.AbstractGitProvider.parseStaticGitContext(repoUrl)
+	if err != nil {
+		return nil, err
+	}
+
+	urlParts := strings.Split(repoUrl, "/")
+	if len(urlParts) < 5 {
+		return nil, fmt.Errorf("Invalid repository URL")
+	}
+
+	// Example URL: https://bitbucket.example.com/projects/<PROJECT_KEY>/repos/<REPO_NAME>
+	var projectKey string
+	for i, part := range urlParts {
```

Comment:
I suggest using regex to extract the project key to make this more readable.

Example:
```golang
re := regexp.MustCompile(`http[s]?://.*/projects/([^/]+)/.*`)
matches := re.FindStringSubmatch(url)

if len(matches) > 1 {
  fmt.Println("PROJECT_KEY:", matches[1])
} else {
  fmt.Println("PROJECT_KEY not found")
}
```

Reply:
done..